MdeModulePkg/Universal/FaultTolerantWriteDxe/FtwMisc.c

```diff
@@ -815,7 +815,10 @@ FtwGetLastWriteHeader (
     //
     // If Offset exceed the FTW work space boudary, return error.
     //
-    if (Offset >= FtwWorkSpaceSize) {
+    // MU_CHANGE [BEGIN] - Prevent buffer overrun

```

Comment:
If this function is receiving `FtwWorkSpaceHeader` as input, I think it should not make any assumptions about previous validation of the buffer contents and check the [`Signature` at the beginning of the structure](https://github.com/microsoft/mu_basecore/blob/release/202311/MdeModulePkg/Include/Guid/SystemNvDataGuid.h#L37), that's why it is there.

Reply:
@makubacki 

as per the comment of this function the last written header may be an empty one

The last write header is the header whose 'complete' state hasn't been set.
  **After all, this header may be a EMPTY header entry for next Allocate.**

Reply:
It is checking the `Complete` field in a `EFI_FAULT_TOLERANT_WRITE_HEADER` which should follow a `EFI_FAULT_TOLERANT_WORKING_BLOCK_HEADER`.

![image](https://github.com/user-attachments/assets/d65b2a9d-d48d-441f-885e-237216b3b388)

In the above image:

- `FtwWorkSpaceHeader` is `WORKING_BLOCK_HEADER`
- `FtwHeader` is `WRITE_HEADER`
- `FtwHeader->NumberOfWrites` specifies number of `WRITE_RECORD`s

That is saying "**this header**" may be an "**EMPTY header entry**". I interpret that to mean the header entry would be present and the signature should be valid.

Have you inspected the code flow to know that the `Signature` of `FtwWorkSpaceHeader` would not be valid? I have not looked in detail at that aspect but it seems strange to start processing write headers and records following an invalid working block header.

Reply:
I have not yet looked the entire flow in detail. I will check on the `Signature of FtwWorkSpaceHeader ` and then we can make the appropriate changes

Reply:
I have checked the flow and it looks like FtwGetLastWriteHeader() is called after InitializeLocalWorkSpaceHeader(). 
This InitializeLocalWorkSpaceHeader() does write 0xFF (FTW_ERASED_BYTE) to EFI_FAULT_TOLERANT_WORKING_BLOCK_HEADER and copies the guid gEdkiiWorkingBlockSignatureGuid

So even in negative case (corrupted or malformed block) we will have valid guid in EFI_FAULT_TOLERANT_WORKING_BLOCK_HEADER and signature verification would PASS
hence I believe we cannot make use the signature check here. Please let me know if you think otherwise.

Reply:
If I understand correctly, it is valuable in the case of hardware failure such as wear out of the containing flash sector where a successful write and readback is expected.

Reply:
Yes, this change will be useful in case such as wear, and tear causes the entire block to be 0s instead of 1s as the current code expects.

However, in any case, the code should not access the beyond `FtwWorkSpaceHeader+FtwWorkSpaceSize`

Reply:
The code must be robust if corrupt data is encountered. It does not really matter the source of the corruption, it could be a hardware failure, or from defective or malicious software.

Reply:
Hi @makubacki,   I have added the signature check. Can you please check now.

I will send the changes to EDK2 forum with all these changes

Reply:
created a PR here https://github.com/tianocore/edk2/pull/6045